### PR TITLE
[Tool] Migrate tests to build in hatch test.

### PR DESCRIPTION
**Pull request by dolaameng**
Migrate from "hatch run test" script to "hatch test" builtin. Also migrate kagglehub off python 3.8- support.

- Hatch 1.10+ supports builtin test by `hatch-test` env. This has several pros compared to our own implementation of tests as scripts. Details in the task. We removed the current test script but keep the docker-hatch functionality.
- Stop support for python 3.7/3.8 so the api/tool usage is much clearer. TODO: migrate python code (b/360894076)

FIX=b/353580550

Tested
- Test all python versions: `hatch test --all`
- Test specific python versions: `hatch test -i py=3.11`
- Integration tests: `hatch test integration_tests -py 3.9`

### CODE REVIEW

**Comment by neshdev** [comment]
Do you mean these scripts?
https://screenshot.googleplex.com/RDAsjfEjQ8SoZq2

**Comment by dolaameng** [comment]
> Do you mean these scripts? https://screenshot.googleplex.com/RDAsjfEjQ8SoZq2

Yes.

File: pyproject.toml [hunk 1]
```diff
@@ -55,6 +55,17 @@ cov = [
   "cov-report",
 ]
 
+[tool.hatch.envs.hatch-test]
+dependencies = [
```
Comment (neshdev) [inline pyproject.toml:59]: Parts of these sections are duplicated. Can we just have a single reference?

File: pyproject.toml [hunk 2]
```diff
@@ -55,6 +55,17 @@ cov = [
...
+  "coverage[toml]>=6.5; python_version >= '3.8'",
+  "pytest",
+  "flask",
+  "flask-jwt-extended",
+]
+
+[[tool.hatch.envs.hatch-test.matrix]]
+python = ["3.7", "3.8", "3.9", "3.10", "3.11"]
```
Comment (neshdev) [inline pyproject.toml:67]: i thought 3.7 does not support hatch test?

File: pyproject.toml (hunk 1 above)
Comment (dolaameng) [reply pyproject.toml:59]: Yes this a good point. We can avoid the duplicates by templating, e.g.,
```
[tool.hatch.envs.hatch-test]
dependencies = [
  "coverage[toml]>=6.5; python_version >= '3.8'",
  "pytest",
  "flask",
  "flask-jwt-extended",
]

[tool.hatch.envs.default]
template = "hatch-test"
```

But this is not supported by earlier hatch version 1.7.0, which is run under python 3.7. Or I lack the knowledge of doing this so please feel free to suggest.

This early hatch version is currently used by `docker-hatch`. If we decide that we don't need it anymore, we can refactor it here.

File: pyproject.toml (hunk 2 above)
Comment (dolaameng) [reply pyproject.toml:67]: Yes if we install hatch under py=3.7, it will install an early version which is like 1.7.0. There are a lot of hatch changes that won't work in this early version. That's another reason why we may consider deprecating docker-hatch some day.

File: Dockerfile [hunk 3]
```diff
@@ -2,7 +2,7 @@ ARG PYTHON_VERSION
 
 FROM python:${PYTHON_VERSION}
 
-RUN pip install hatch==1.9.4
```
Comment (rosbo) [inline Dockerfile:5]: We want to keep this version pinned to ensure everyone testing with `docker-hatch` have the same exact environment to prevent issues like "it runs in mine but not yours". The docker image is not rebuilt automatically at every run. If nothing changes in the Dockerfile, it won't get rebuilt. This means some people could have a older version of hatch and get inconsistent behaviors.

tl;dr: If you need a newer version of `hatch`, pin it to a newer version.

File: README.md [hunk 4]
```diff
@@ -153,11 +153,14 @@ Follow these [instructions](https://hatch.pypa.io/latest/install/) to install it
 ### Tests
 
 ```sh
-# Run all tests
-hatch run test
```
Comment (rosbo) [inline README.md:157]: I would get ride of the existing `test` script. No need to keep them around. We should all switch to use the new commands.

**Review by rosbo — CHANGES_REQUESTED** [review]
Thanks for working on this!

Can you also make sure to update our cicd pipeline to use the new commands: https://github.com/Kaggle/kagglehub/blob/c0de979194b2e9a623644cf7021c47953ecc0ff6/tools/cicd/cloudbuild.yaml#L35-L36

And for the integration tests: https://github.com/Kaggle/kagglehub/blob/c0de979194b2e9a623644cf7021c47953ecc0ff6/tools/cicd/integration-tests.yaml#L46

File: README.md (hunk 4 above)
Comment (dolaameng) [reply README.md:157]: Do we still want to keep `docker-hatch`, which in my opinion is still useful at least during the migration?

If so, we will need to keep the current `test script`, because for `docker-hatch -v 3.7`, it installs an older version of hatch, which doesn't support new `hatch test`.

File: Dockerfile (hunk 3 above)
Comment (dolaameng) [reply Dockerfile:5]: It seems that hatch >= 1.7.0 will now require python3.8+, which means if we want to install hatch under py3.7, hatch 1.9.4 won't work anymore. I wonder how you got this work for py 3.7 actually.

File: pyproject.toml (hunk 1 above)
Comment (rosbo) [reply pyproject.toml:59]: `docker-hatch` is simply a way to run any hatch command inside a consistent docker image. 

No need for people to install python & hatch locally...

`docker-hatch` also allows you to quickly run a command for various python version. This is useful to help debug commands when you see them failing on ci/cd for a specific version of python (e.g. ./docker-hatch -v 3.9 run lint:all): https://github.com/Kaggle/kagglehub/blob/c0de979194b2e9a623644cf7021c47953ecc0ff6/docker-hatch#L25-L29

File: Dockerfile (hunk 3 above)
Comment (rosbo) [reply Dockerfile:5]: Can you add the link? 

You can easily test hatch with different version of python using `docker-hatch` and see what works and what doesn't.

File: README.md (hunk 4 above)
Comment (dolaameng) [reply README.md:157]: Yes the challenge is, hatch that works with py3.7 (hatch <= 1.10.0) won't have builtin `test`, so we will still need to run the `script test` for python 3.7, if we keep docker-hatch?

File: pyproject.toml (hunk 1 above)
Comment (dolaameng) [reply pyproject.toml:59]: @rosbo So I understand currently `docker-hatch` works as this:
1. We first install the python with specific `PYTHON_VERSION`.
2. We install hatch that works with the current python version.

The challenge of step 2 is: 
1. we will need to install different hatch versions based on whether py <= 3.7 or not: hatch==1.7.0 for py<=3.7 and latest hatch for py > 3.7.
2. When we install different hatch version, `hatch test` doesn't work out of box for hatch <= 1.10.0, so we will still need `script test`. 

It seems we have different options:
1. Removing docker-hatch`, and use the latest hatch version (e.g., 1.12.0). Testing and linting for different python versions are supported this way (see PR summary above).
2. Having conditions for hatch installation, and run different commands `hatch test` vs `hatch run test` for python 3.7 vs later, e.g.,

```
ARG PYTHON_VERSION

FROM python:${PYTHON_VERSION}

RUN if [ "$(printf '%s\n' "3.7" "${PYTHON_VERSION}" | sort -V | head -n 1)" != "${PYTHON_VERSION}" ]; then \
        pip install "hatch==1.7.0"; \
    else \
        pip install "hatch==1.12.0"; \
    fi
```

What do you think in this case?

File: Dockerfile (hunk 3 above)
Comment (dolaameng) [reply Dockerfile:5]: Yes I think https://pypi.org/project/hatch/1.7.0/ is the latest hatch version that has python 3.7 has rest in peace (https://devguide.python.org/versions/). I wonder what Keras team thinks about this, maintaining python 3.7 seems to be a pain in long run.

**Comment by dolaameng** [comment]
> Thanks for working on this!
> 
> Can you also make sure to update our cicd pipeline to use the new commands:
> 
> https://github.com/Kaggle/kagglehub/blob/c0de979194b2e9a623644cf7021c47953ecc0ff6/tools/cicd/cloudbuild.yaml#L35-L36
> 
> And for the integration tests:
> 
> https://github.com/Kaggle/kagglehub/blob/c0de979194b2e9a623644cf7021c47953ecc0ff6/tools/cicd/integration-tests.yaml#L46

Sorry I think the same challenge here. If we still use the docker image built for python 3.7, we are forced to use an early hatch version, which doesn't support "builtin test".

File: Dockerfile (hunk 3 above)
Comment (rosbo) [reply Dockerfile:5]: Let's ask the Keras team, it would simplify things greatly if we don't have to have this weird dual support of different hatch versions and command. 

I will create a thread with them and tag you.

Comment (rosbo) [reply Dockerfile:5]: I created a thread here: https://chat.google.com/room/AAAAZAWC3mc/WY3msYqOvUA/WY3msYqOvUA?cls=10

Comment (rosbo) [reply Dockerfile:5]: Looks like they only support 3.9+ after all. Let's drop support for 3.7 & 3.8.

Comment (dolaameng) [reply Dockerfile:5]: Cool! Thank you for taking actions! Let me work on the migration for hatch in this PR, and maybe we can create other PRs to update source code?

Comment (rosbo) [reply Dockerfile:5]: Here is a PR to remove our GitHub check (GCB trigger) for Python 3.7: https://github.com/Kaggle/kaggleazure/pull/31120

Note: We didn't have a trigger for Python 3.8.

Comment (rosbo) [reply Dockerfile:5]: SG, to use separate PRs to update source code. Thanks for tackling this diligently :)

File: README.md [hunk 5]
```diff
@@ -153,11 +153,14 @@ Follow these [instructions](https://hatch.pypa.io/latest/install/) to install it
...
+# Run all tests for all Python versions.
+hatch test --all
```
Comment (rosbo) [inline README.md:157]: Does `hatch test` work and pick a default Python version?

File: tools/cicd/cloudbuild.yaml [hunk 6]
```diff
@@ -32,7 +32,6 @@ steps:
 - name: us-docker.pkg.dev/$PROJECT_ID/tools/hatch:${_PYTHON_VERSION}
   id: tests
   args:
-  - run
```
Comment (dolaameng) [inline tools/cicd/cloudbuild.yaml:35]: @rosbo When I changed this line, it caused failures of cicd with error that "test" is not supported by hatch. I wonder if our `Google Cloud Build / kagglehub-branch-3-12-0 (kaggle-cicd)` is using a different docker image other than the Dockerfile defined here? How should I debug this?

Comment (rosbo) [reply tools/cicd/cloudbuild.yaml:35]: Yes, you will need to update the `cicd` Dockerfile: https://github.com/Kaggle/kagglehub/blob/main/tools/cicd/Dockerfile

File: README.md (hunk 5 above)
Comment (dolaameng) [reply README.md:157]: You mean the default python version? I assume it's using current python version which installed hatch.

Comment (rosbo) [reply README.md:157]: I would leave it documented here. This is likely the command most people will use.

File: pyproject.toml [hunk 7]
```diff
@@ -34,29 +34,22 @@ dependencies = [
...
-cov = [
-  "test-cov",
-  "cov-report",
-]
+[[tool.hatch.envs.hatch-test.matrix]]
+python = ["3.9", "3.10", "3.11"]
 
 [[tool.hatch.envs.all.matrix]]
-python = ["3.7", "3.8", "3.9", "3.10", "3.11"]
+python = ["3.9", "3.10", "3.11"]
```
Comment (rosbo) [inline pyproject.toml:49]: We should also add 3.12 to match our GCB triggers: https://github.com/Kaggle/kaggleazure/blob/aa1de82d6f1dc71c5715c40b6327053ead4648b7/configs/cicd/cloud_builds/main.tf#L53-L58

File: README.md (hunk 5 above)
Comment (dolaameng) [reply README.md:160]: Good point. Done.

File: pyproject.toml (hunk 7 above)
Comment (dolaameng) [reply pyproject.toml]: Done.

File: README.md [hunk 8]
```diff
@@ -153,11 +153,17 @@ Follow these [instructions](https://hatch.pypa.io/latest/install/) to install it
 ### Tests
 
 ```sh
-# Run all tests
-hatch run test
+# Run all tests for current Python version.
+hatch test
```
Comment (rosbo) [inline README.md:157]: Just to confirm, this only runs the unit tests correct? (i.e. doesn't include the integration tests)

**Review by rosbo — APPROVED** [review]
Thanks for spearheading this effort!

File: README.md (hunk 8 above)
Comment (dolaameng) [reply README.md:157]: Correct! By default the "args" to `hatch test` is the tests folder (https://hatch.pypa.io/1.12/config/internal/testing/#default-arguments). We need to explicitly pass `hatch test integration_tests` for integration tests.
